riscv32: move memory definitions to dts

mateusz-holenko opened the pull request:
Currently memory for riscv32 is defined in Kconfig. Move those
definitions to DTS and use defines generated from it in linker script.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

galak reviewed (changes requested):
Please break this up into multiple comments for each step of the changes involved here.

galak commented on dts/bindings/sram/sifive,dtim0.yaml:
```diff
@@ -0,0 +1,33 @@
+      generation: define
+      category: required
+
+    label:
```
remove label, not used or needed for dtim0.

galak commented on dts/riscv32/riscv32-fe310.dtsi:
```diff
@@ -35,6 +35,10 @@
 		compatible = "SiFive,FE310G-0002-Z0-soc", "fe310-soc",
 			"sifive-soc", "simple-bus";
 		ranges;
+		flash0: flash@20400000 {
```
What is this flash?  I thought the fe310 had an off chip qspi or something like that?

nategraff-sifive replied:
I think this flash refers to the region of the QSPI that's reserved for the application.

nategraff-sifive replied:
This might be better served by recognizing the memory addressed by the spi0 node as the FLASH_BASE_ADDRESS and then configuring 0x400000 as the FLASH_LOAD_OFFSET. Which leads me back to my question of whether we should use CONFIG_FLASH_BASE_ADDRESS on RISC-V like other arches.

galak replied:
This is still not correct, look at `boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts`.  

Take a look at:
```
&flexspi0 {
        reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
        is25wp064: is25wp064@0 {
                compatible = "issi,is25wp064", "jedec,spi-nor";
                reg = <0>;
                status = "ok";
        };
};
```

mateusz-holenko replied:
You are right, we were mistaken.
FE-310 does not have a built-in flash, it's a part of HiFive board.

I'm thinking of splitting the dts into two files: 

- one for the SoC itself (without flash) 
- and second for the board (it would include SoC's dts and add flash).

Is this what you meant by referencing `mimxrt1060_evk.dts`?

mateusz-holenko replied:
Since there had already been a dts for HiFive1 board, I just moved flash definition there and added it under `spi0` (just like in the `flexspi0` example).

mateusz-holenko replied:
I had to duplicate the flash definiton in `qemu_riscv32.dts` as it is based on `riscv32-fe310.dtsi`.

mateusz-holenko replied:
@galak Is this solution what you expected?

galak commented on soc/riscv32/openisa_rv32m1/soc.h:
```diff
@@ -105,7 +107,16 @@ void soc_interrupt_init(void);
+#define RISCV_ROM_SIZE DT_SOC_NV_FLASH_0_SIZE
+#define RISCV_RAM_BASE DT_MMIO_SRAM_20000000_BASE_ADDRESS
+#define RISCV_RAM_SIZE DT_MMIO_SRAM_20000000_SIZE
+#else
```
maybe:

```
#elif (CONFIG_SOC_OPENISA_RV32M1_ZERO_RISCY)`
...
#else 
#error not soc defined
#endif 
```

mateusz-holenko replied:
done

galak reviewed (dismissed):
Some additional comments.

nategraff-sifive commented on dts/bindings/sram/sifive,dtim0.yaml:
```diff
@@ -0,0 +1,33 @@
+version: 0.1
+
+description: >
+    This binding gives description of the SiFive Data Tighly Integrated Memory
```
Minor change: "This binding describes the SiFive Data Tightly-Integrated Memory"

mateusz-holenko replied:
Fixed.

nategraff-sifive reviewed (commented):
@galak I've noticed that other architectures use the FLASH_BASE_ADDRESS config option. Would now be a good time to get RISC-V on that standard? How does that relate to this PR?

fkokosinski committed:
dts: bindings: sram: add SiFive dtim0 bindings

Add bindings for SiFive Data Tightly-Integrated Memory.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

fkokosinski committed:
dts: riscv32: microsemi-miv: add flash and sram

Add flash and SRAM to the Microsemi MiV device tree.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

fkokosinski committed:
riscv32: use device tree defines in linker

Delete memory-related configs from defconfig and use device tree based
macros in general riscv32 linker script instead of Kconfig ones.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

MaureenHelm commented on CODEOWNERS:
```diff
@@ -139,6 +139,7 @@
 /dts/bindings/can/                        @alexanderwachter
 /dts/bindings/*/nordic*                   @anangl
 /dts/bindings/*/nxp*                      @MaureenHelm
+/dts/bindings/sram/sifive,dtim0.yaml      @mateusz-holenko @kgugala @pgielda @nategraff-sifive
```
Make this more generic like the two previous entries. Something like:
```
/dts/bindings/*/sifive*      @mateusz-holenko @kgugala @pgielda @nategraff-sifive
```

mateusz-holenko replied:
Done

MaureenHelm commented on dts/riscv32/rv32m1_ri5cy.dtsi:
```diff
@@ -41,6 +41,21 @@
+	/*
+	 * Assumes ARM cores are disabled. Please note that these values are
+	 * a subject to change if SOC_OPEN_ISA_RV32M1_RI5CY_ZERO
+	 * is enabled in config.
```
In moving memory definitions to dts, these values shouldn't change. We should instead define multiple flash regions, with the appropriate region being chosen based on the core selected.

I have a PR to enable the zero-riscy core in #15542.

henrikbrixandersen replied:
I agree. Zephyr does not currently support either of the ARM cores (or zero-risky, for that matter), but hopefully support for these will land in the tree soon.

mateusz-holenko replied:
I splitted the device tree information into three files: base (the common one) and two specific to riscy and zero_riscy.

mateusz-holenko replied:
@MaureenHelm Is the current solution ok?

MaureenHelm replied:
No, please see the last two commits in #15943

mateusz-holenko replied:
I see that your PR is merged to master, so I rebased onto it and dropped my changes.

mateusz-holenko replied:
@MaureenHelm I guess this closes the discussion, please confirm.

MaureenHelm replied:
Yes, thank you

mateusz-holenko commented:
I addressed all the requested changes. Is there anything else I should do?

nategraff-sifive commented on dts/bindings/sram/sifive,dtim0.yaml:
```diff
@@ -0,0 +1,27 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+---
+title: Data Tighly Integrated Memory
```
Sorry to nitpick. s/Tighly/Tightly/

mateusz-holenko replied:
Fixed, thanks for spotting it.

nategraff-sifive commented on soc/riscv32/riscv-privilege/sifive-freedom/soc.h:
```diff
@@ -40,7 +41,11 @@
+
+/* Flash defined variables for linker */
+#define RISCV_ROM_BASE		DT_SOC_NV_FLASH_0_BASE_ADDRESS
+#define RISCV_ROM_SIZE		DT_SOC_NV_FLASH_0_SIZE
```
Instead of hard-coding the device we expect for RAM and ROM, we should use the chosen node in the devicetree.
```
chosen {
		zephyr,sram = &dtim;
		zephyr,flash = &flash0;
};
```

nategraff-sifive replied:
Especially since other SiFive SoCs won't necessarily have a DTIM or soc-nv-flash.

mateusz-holenko replied:
I changed it to use `DT_SRAM` and `DT_FLASH` symbols. Is it ok to assume that both ROM and RAM is present or should it be ifdefed somehow (e.g., by using `DT_CHOSEN_ZEPHYR_FLASH`)?

nategraff-sifive replied:
Awesome, perfect!

I'm *pretty sure* this is sufficient, since [include/arch/riscv32/common/linker.ld](https://github.com/zephyrproject-rtos/zephyr/tree/master/include/arch/riscv32/common/linker.ld) ignores `CONFIG_RISCV_ROM_BASE_ADDR` if `CONFIG_XIP=n`.

fkokosinski committed:
soc: riscv32: sifive-freedom: soc.h: use defines from device tree

Use values generated from the device tree in RISCV_ROM_BASE,
RISCV_ROM_SIZE, RISCV_RAM_BASE, RISCV_RAM_SIZE macros.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

fkokosinski committed:
dts: riscv32: hifive1: add flash

Add flash to the SiFive HiFive1 device tree.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

mateusz-holenko committed:
dts: riscv32: hifive1: configure sram

Choose dtim as sram in the SiFive HiFive1 device tree.

Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

mateusz-holenko committed:
dts: riscv32: qemu_riscv32: configure sram

Choose dtim as sram in QEMU RISC-V 32 device tree.

Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

mateusz-holenko committed:
dts: riscv32: qemu_riscv32: add flash

Add flash to the QEMU RISC-V 32 device tree.

Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

nategraff-sifive reviewed (approved):
The common RISC-V and SiFive changes LGTM.

Someone with more familiarity with the MiV and Vega platforms should still approve those changes.

carlescufi commented:
@galak @MaureenHelm can you take another look?

MaureenHelm reviewed (changes requested):
https://github.com/zephyrproject-rtos/zephyr/pull/15418#discussion_r281316445

fkokosinski committed:
soc: riscv32: miv: soc.h: use defines from device tree

Use values generated from the device tree in RISCV_ROM_BASE,
RISCV_ROM_SIZE, RISCV_RAM_BASE, RISCV_RAM_SIZE macros.

Signed-off-by: Filip Kokosinski <fkokosinski@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

mateusz-holenko committed:
dts: m2gl025_miv: configure flash and sram

Choose sram0 as sram and flash0 as flash in m2gl025_miv device tree.

Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

MaureenHelm commented on boards/riscv32/hifive1/hifive1.dts:
```diff
@@ -52,6 +53,13 @@
+	flash0: flash@20400000 {
+		compatible = "soc-nv-flash";
+		reg = <0x20400000 0xc00000>;
+	};
```
@galak please confirm, but I think this should be (assuming spi chip select 0):
```
	reg = <0x10014000 0x1000>, <0x20400000 0xc00000>;
	flash0: flash@0 {
		compatible = "soc-nv-flash";
		reg = <0>;
	};
```

MaureenHelm replied:
This still needs to be fixed

mateusz-holenko replied:
I fixed this and the analogous dts for riscv32 QEMU platform.

MaureenHelm replied:
@galak which is the preferred syntax?
```reg = <0x10014000 0x1000>, <0x20400000 0xc00000>;```
or
```reg = <0x10014000 0x1000 0x20400000 0xc00000>;```

mateusz-holenko replied:
@MaureenHelm 

I used the same notation as in the original [spi0](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/riscv32/riscv32-fe310.dtsi#L181) node just be consistent.

mateusz-holenko commented:
I had to add a new commit modyfing dts parsing script in python, as after rebasing onto the newest master it crashed when handling nested `flash0` node in hifive1 platform.

mgielda commented:
@MaureenHelm I think we're good to go. Can you approve and merge?

MaureenHelm commented on scripts/dts/extract/globals.py:
```diff
@@ -294,7 +294,8 @@ def get_binding(node_path):
             bus = parent_binding['child']['bus']
-            return bus_bindings[bus][compat]
+            if bus in bus_bindings and compat in bus_bindings[bus]:
+                return bus_bindings[bus][compat]
```
Why not just add the bus binding?

henrikbrixandersen replied:
Not really sure why this was needed?

mateusz-holenko replied:
@henrikbrixandersen This was not needed until I rebased onto [this commmit](https://github.com/zephyrproject-rtos/zephyr/commit/f9dd53624c578be51fbb91ddf2938df55a0c6966).

With the new version of `globals.py`, sanitycheck fails with the following error:

```
Zephyr version: 1.14.99
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.7.3", minimum required is "3.4") 
-- Selected BOARD hifive1
-- Found west: /usr/local/bin/west (found suitable version "0.5.8", minimum required is "0.5.6")
-- Loading litex-zephyr/boards/riscv32/hifive1/hifive1.dts as base
-- Overlaying litex-zephyr/dts/common/common.dts
Traceback (most recent call last):
  File "litex-zephyr/scripts/dts/extract_dts_includes.py", line 517, in <module>
    main()
  File "litex-zephyr/scripts/dts/extract_dts_includes.py", line 503, in main
    generate_defines()
  File "litex-zephyr/scripts/dts/extract_dts_includes.py", line 446, in generate_defines
    generate_node_defines(node_path)
  File "litex-zephyr/scripts/dts/extract_dts_includes.py", line 111, in generate_node_defines
    generate_bus_defines(node_path)
  File "litex-zephyr/scripts/dts/extract_dts_includes.py", line 140, in generate_bus_defines
    binding = get_binding(node_path)
  File "litex-zephyr/scripts/dts/extract/globals.py", line 297, in get_binding
    return bus_bindings[bus][compat]
KeyError: 'soc-nv-flash'
CMake Error at litex-zephyr/cmake/dts.cmake:174 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  litex-zephyr/cmake/app/boilerplate.cmake:403 (include)
  CMakeLists.txt:4 (include)


-- Configuring incomplete, errors occurred!
```

Having in mind that:

-  it worked before the rebase, 
- seeing [a comment](https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/dts/extract/globals.py#L299) and 
- seeing that all previous accesses to the dictionaries are protected, 

I assumed that the code should fall-back to [the final return](https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/dts/extract/globals.py#L300) and not throw an exception on a dictionary access.

mateusz-holenko replied:
@MaureenHelm I'm affraid I don't follow. I'm not very fluent in the dts-extraction code, so if there is a better place where the fix should go or the DTS itself should be reworked, please provide me with more details.

henrikbrixandersen replied:
@mateusz-holenko Ahh, now I see it. It is due to you using `compatible = "soc-nv-flash"` for something, that's an external SPI flash, so the DTS YAML binding does not include the SPI device bus setup (as defined in `spi-device.yaml `).

I am not familiar with what kind of flash is present on the HiFive1. Can you use `compatible = "jedec,spi-nor"` instead?

mateusz-holenko replied:
@henrikbrixandersen Thanks for the hint.

According to [this doc](https://sifive.cdn.prismic.io/sifive%2F7503d08c-5f6f-432e-a721-bb02f75cb5e7_hifive1-a01-bom.pdf) (position 43), this is [IS25LP128-JBLE](https://eu.mouser.com/datasheet/2/198/IS25LP032-064-128-463542.pdf) flash. [Linux driver](https://github.com/torvalds/linux/blob/master/drivers/mtd/spi-nor/spi-nor.c#L1826) uses `0x9d6019` as JEDEC ID.

My proposition is to describe it as:
```
      flash0: flash@0 {
              compatible = "issi,is25lp128", "jedec,spi-nor";
              jedec-id = <0x96 0x60 0x18>;
              reg = <0>;
      };
```

With this shape of the DTS my change in `globals.py` is not needed anymore.

mateusz-holenko replied:
Ok, I did that. Does this work for you @MaureenHelm?

henrikbrixandersen commented on include/arch/riscv32/common/linker.ld:
```diff
@@ -36,9 +36,9 @@
+    ROM (rx)  : ORIGIN = RISCV_ROM_BASE, LENGTH = RISCV_ROM_SIZE
 #endif
-    RAM (rwx) : ORIGIN = CONFIG_RISCV_RAM_BASE_ADDR, LENGTH = RISCV_RAM_SIZE
+    RAM (rwx) : ORIGIN = RISCV_RAM_BASE, LENGTH = RISCV_RAM_SIZE
```
Why do we need to come up with new, architecture-specific indirection variables for RAM and ROM base and size? Why not just rely on `DT_SRAM_BASE_ADDRESS` + `DT_RAM_SIZE` and friends directly?

mateusz-holenko replied:
`RISCV_RAM_BASE`/`RISCV_RAM_SIZE` are not new things, they are currently defined in [openisa_rv32m1](https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/riscv32/openisa_rv32m1/soc.h#L108), [sifive-freedom](https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/riscv32/riscv-privilege/sifive-freedom/soc.h#L43), [miv](https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/riscv32/riscv-privilege/miv/soc.h#L57) and referenced in: [libc-hooks.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/libc/newlib/libc-hooks.c#L46).

We just used them for `RAM` and, here you are totally right, added the analugous set of defines for `ROM`.

I agree that using `DT_` directly reduces the indirection level.
I can rework `linker.ld` and drop `RISCV_ROM_BASE`/`RISCV_RAM_BASE`, but I would leave `RISCV_RAM_BASE`/`RISCV_RAM_SIZE` as it is as reworking `libc` is outside the scope of this PR.

mgielda commented:
Hi Maureen, this is blocking a lot of other work on our side. If there is some discussion "something that works well" vs "something that works just as well", can we move it to a separate issue to be resolved independently? This has been going on forever.

mateusz-holenko commented:
I see that shippable failed on [one qemu test](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/41439/5/tests), but when I run this test on **master** on my desktop it failed as well. Is this  a known issue?

mgielda commented:
Thanks! We will monitor for minor improvements we can do on top of that. Meanwhile we will continue with adding new platforms.
